dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventCategoryOptValidationHook.java

```diff
@@ -0,0 +1,227 @@
+import org.hisp.dhis.category.CategoryOptionCombo;
+import org.hisp.dhis.common.IllegalQueryException;
+import org.hisp.dhis.commons.util.TextUtils;
+import org.hisp.dhis.i18n.I18nFormat;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.preheat.PreheatHelper;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.util.DateUtils;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class EventCategoryOptValidationHook
+    extends AbstractTrackerValidationHook
+{
+
+    @Override
+    public int getOrder()
+    {
+        return 303;
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerBundle bundle )
+    {
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, this.getClass() );
+        User actingUser = bundle.getPreheat().getUser();
+
+        for ( Event event : bundle.getEvents() )
+        {
+            reporter.increment( event );
+
+            ProgramStage programStage = PreheatHelper.getProgramStage( bundle, event.getProgramStage() );
+            ProgramInstance programInstance = PreheatHelper.getProgramInstance( bundle, event.getEnrollment() );
+            TrackedEntityInstance trackedEntityInstance = PreheatHelper
+                .getTrackedEntityInstance( bundle, event.getTrackedEntityInstance() );
+            Program program = PreheatHelper.getProgram( bundle, event.getProgram() );
+
+            if ( program == null )
+            {
+                continue;
+            }
+
+            programStage = (programStage == null && program.isWithoutRegistration())
+                ? program.getProgramStageByStage( 1 ) : programStage;
+            if ( programStage == null )
+            {
+                continue;
+            }
+
+            programInstance = getProgramInstance( actingUser, programInstance, trackedEntityInstance, program );
+            program = programInstance.getProgram();
+
+            validateCategoryOptionCombo( bundle, reporter, actingUser, event, program );
+        }
+
+        return reporter.getReportList();
+    }
+
+    private void validateCategoryOptionCombo( TrackerBundle bundle, ValidationErrorReporter errorReporter
+        , User user, Event event, Program program )
+    {
+
+        // NOTE: Morten H. & Stian. Abyot : How do we solve this in the new importer?
```

Comment:
I am not too familiar with coc in tracker, so I will leave this to @abyot and @mortenoh

Reply:
There are 3 things to validate when it comes of coc.

- read/write access, but this needs to be done through categoryOption (not coc itself). User needs to have access to all categorypOptions of the coc. If one categoryOption, fails then the whole thing fails.
- the organisationUnit (event's organisationUnit) needs to have access to all categoryOptions of the coc
- event's date should fall within the open dates of all category options of the coc.

1.